File: runtime/lib/phases/verify.ts
import { createHash } from 'node:crypto';
import { readFile } from 'node:fs/promises';
import type { Artifact, Integrity } from '@torba/core';
import { integrityHashes } from '@torba/core';

async function checkHash(
  path: string,
  algo: 'sha1' | 'sha256' | 'md5',
  expected: string,
): Promise<boolean> {
  const hash = createHash(algo);
  hash.update(await readFile(path));
  return hash.digest('hex') === expected;
}

export function verifyIntegrity(
  path: string,
  integrity: Integrity | undefined,
): Promise<boolean> {
  const entries = integrityHashes(integrity);
  if (entries.length === 0) return Promise.resolve(true);
  // Each entry names exactly one algorithm; the file passes if any entry
  // matches. A read error (e.g. missing file) resolves to `false`.
  return Promise.all(
    entries.map((e) => {
      if ('sha1' in e) return checkHash(path, 'sha1', e.sha1);
      if ('sha256' in e) return checkHash(path, 'sha256', e.sha256);
      return checkHash(path, 'md5', e.md5);
    }),
  ).then(
    (results) => results.some(Boolean),
    () => false,
  );
}

export async function verifyAll(
  tasks: { finalPath: string; artifact: Artifact }[],
): Promise<string[]> {
  const results = await Promise.all(
    tasks.map(async ({ finalPath, artifact }) => {
      const ok = await verifyIntegrity(finalPath, artifact.integrity);
      return ok ? null : finalPath;
    }),
  );
  return results.filter((p): p is string => p !== null);
}

File: crates/torba-runtime/src/phases/verify.rs
use md5::Md5;
use sha1::Sha1;
use sha2::{Digest, Sha256};
use tokio::fs;
use torba_core::{Artifact, HashEntry, Integrity};

fn check_hash(data: &[u8], entry: &HashEntry) -> bool {
    let computed = match entry {
        HashEntry::Sha1 { .. } => {
            let mut h = Sha1::new();
            h.update(data);
            hex::encode(h.finalize())
        }
        HashEntry::Sha256 { .. } => {
            let mut h = Sha256::new();
            h.update(data);
            hex::encode(h.finalize())
        }
        HashEntry::Md5 { .. } => {
            let mut h = Md5::new();
            h.update(data);
            hex::encode(h.finalize())
        }
    };
    computed.eq_ignore_ascii_case(entry.hex())
}

pub async fn verify_integrity(path: &str, integrity: Option<&Integrity>) -> bool {
    let entries = integrity.map(Integrity::entries).unwrap_or(&[]);
    if entries.is_empty() {
        return true;
    }
    let Ok(data) = fs::read(path).await else {
        return false;
    };
    entries.iter().any(|e| check_hash(&data, e))
}

pub async fn verify_all<'a>(
    tasks: impl IntoIterator<Item = (&'a str, &'a Artifact)>,
) -> Vec<String> {
    let mut failures = Vec::new();
    for (path, artifact) in tasks {
        if !verify_integrity(path, artifact.integrity.as_ref()).await {
            failures.push(path.to_owned());
        }
    }
    failures
}
